### PR TITLE
DAT-3391 maintenance script for retrieving templates

### PR DESCRIPTION
**Pull request by dianafa**
https://wikia-inc.atlassian.net/browse/DAT-3391

Script collects all templates from given `SERVER_ID` which are used in content namespace at least once and sends their `cityId` and `pageId` to rabbit queue with message prefix `articledi` and vhost `tests` (needs to be changed for testing in `PipelineConnectionBase` class).

A single generated message payload looks like:
`{"cityId":"831","pageId":"76533"}`

### CODE REVIEW

File: extensions/wikia/TemplateListProvider/TemplateListProviderController.class.php [hunk 1]
```diff
@@ -0,0 +1,44 @@
+<?php
```
Comment (idradm) [inline extensions/wikia/TemplateListProvider/TemplateListProviderController.class.php]: What's the purpose of this controller? Why do we need it?

File: maintenance/wikia/topTemplatesFromWiki.php [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+/**
+ *
+ * @ingroup Maintenance
+ */
+
+require_once( dirname( __FILE__ ) . '../../Maintenance.php' );
+
+class topTemplatesFromWiki extends Maintenance {
+	const TEMPLATE_MESSAGE_PREFIX = 'articledi';
```
Comment (idradm) [inline maintenance/wikia/topTemplatesFromWiki.php]: Did you mean "articlejedi"?

**Comment by idradm** [comment]
What about sql query performance? @SebastianMarzjan who should take a look at that?

File: maintenance/wikia/topTemplatesFromWiki.php [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+			->GROUP_BY('tl_title')
+			->HAVING( 'COUNT(*)' )->GREATER_THAN( 0 )
+			->ORDER_BY('COUNT(*)')->DESC()
+			->runLoop( $db, function ( &$pages, $row ) {
+				$pages[] = [
+					'page_id' => $row->temp_id,
+					'title' => $row->tl_title
+				];
+			} );
+
```
Comment (dianafa) [inline maintenance/wikia/topTemplatesFromWiki.php:52]: `explain SELECT p2.page_id as temp_id, tl_title, COUNT(*)  FROM page AS p  INNER JOIN templatelinks AS t  ON  t.tl_from =  p.page_id  INNER JOIN page AS p2  ON  p2.page_title =  t.tl_title  WHERE p.page_namespace = '0'  AND p2.page_namespace = '10' GROUP BY tl_title HAVING COUNT(*) > 0 ORDER BY COUNT(*) DESC;
+----+-------------+-------+--------+----------------------+------------+---------+-------------------------+-------+----------------------------------------------+
| id | select_type | table | type   | possible_keys        | key        | key_len | ref                     | rows  | Extra                                        |
+----+-------------+-------+--------+----------------------+------------+---------+-------------------------+-------+----------------------------------------------+
|  1 | SIMPLE      | t     | index  | tl_from,tl_namespace | tl_from    | 265     | NULL                    | 63248 | Using index; Using temporary; Using filesort |
|  1 | SIMPLE      | p2    | eq_ref | name_title           | name_title | 261     | const,muppet.t.tl_title |     1 | Using index                                  |
|  1 | SIMPLE      | p     | eq_ref | PRIMARY,name_title   | PRIMARY    | 4       | muppet.t.tl_from        |     1 | Using where                                  |
+----+-------------+-------+--------+----------------------+------------+---------+-------------------------+-------+----------------------------------------------+`

(explain in another form sent via different way)

File: maintenance/wikia/topTemplatesFromWiki.php [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+	 *
+	 * @return bool|mixed
+	 * @throws \Exception
+	 * @throws \FluentSql\Exception\SqlException
+	 */
+	protected function getTemplatesFromWiki() {
+		$db = wfGetDB( DB_SLAVE );
+
+		$pages = ( new \WikiaSQL() )
+			->SELECT('page_id','page_title')
```
Comment (dianafa) [inline maintenance/wikia/topTemplatesFromWiki.php:43]: `explain select page_id, page_title from page where page_namespace = 10;`

+----+-------------+-------+------+---------------+------------+---------+-------+------+-------------+
| id | select_type | table | type | possible_keys | key        | key_len | ref   | rows | Extra       |
+----+-------------+-------+------+---------------+------------+---------+-------+------+-------------+
|  1 | SIMPLE      | page  | ref  | name_title    | name_title | 4       | const |  172 | Using index |
+----+-------------+-------+------+---------------+------------+---------+-------+------+-------------+

Comment (sebastianmarzjan) [inline maintenance/wikia/topTemplatesFromWiki.php:43]: :+1:

File: maintenance/wikia/topTemplatesFromWiki.php [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+	 * Set script options
+	 */
+	public function __construct() {
+		parent::__construct();
+		$this->mDescription = 'allTemplatesFromWiki';
+	}
+
+	public function execute() {
+		$data = $this->getTemplatesFromWiki();
+		$this->pushDataToRabbit( $data );
```
Comment (sebastianmarzjan) [inline maintenance/wikia/topTemplatesFromWiki.php]: Do you want to have output also before getting templates and informing about push start?

File: maintenance/wikia/topTemplatesFromWiki.php [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+
+/**
+ *
+ * @ingroup Maintenance
+ */
+
+require_once( dirname( __FILE__ ) . '../../Maintenance.php' );
+
+class allTemplatesFromWiki extends Maintenance {
+	const TEMPLATE_MESSAGE_PREFIX = 'articledi';
```
Comment (sebastianmarzjan) [inline maintenance/wikia/topTemplatesFromWiki.php]: This name looks temporary, should it stay like that?

File: maintenance/wikia/topTemplatesFromWiki.php [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+		global $wgCityId;
+		$msg = new stdClass();
+		$msg->cityId = $wgCityId;
+		$msg->args = new stdClass();
+		foreach ( $data as $template ) {
+			$msg->pageId = (int)$template[ 'page_id' ];
+
+			try {
+				self::getPipeline()
+					->publish( implode( '.', [ self::TEMPLATE_MESSAGE_PREFIX ] ), $msg );
```
Comment (sebastianmarzjan) [inline maintenance/wikia/topTemplatesFromWiki.php:69]: Do you want to use the 'new' naming convention?

File: maintenance/wikia/topTemplatesFromWiki.php (hunk 6 above)
Comment (dianafa) [inline maintenance/wikia/topTemplatesFromWiki.php]: I wanted to leave it temporary until we're sure about the message prefix

**Comment by sebastianmarzjan** [comment]
LGTM

**Pull request merged**